Improves Giftwrapping

emmanuelbassil opened the pull request:
🆑 emmanuelbassil
tweak: Now able to wrap crates/closets with gift wrapping paper; and to wrap humans with regular parcel wrapping paper. Both have the same behavior.
tweak: Can now destination tag and label gifts.
tweak: You now require a sharp item to open any wrapped parcel
tweak: No more instant wrapping. Time delay scales up the bigger the item you're wrapping is. Also uses more wrap depending on item size.
tweak: Can now clip foamclogs with any sharp item, not only wirecutters or the scalpel. 
bugfix: Instant Tactical combat gift-wrapping is no longer possible. You need a strong grab and a delay to gift-wrap someone else, like handcuffs.
/🆑 

Rewrote all of gift and giftwrap and made them work exactly like parcel wrap.
Gifts are now a resprited/renamed version of the smalldelivery item; with the renaming/respriting handled in Initialize(). As a result, destroyed obj/item/gift completely.
Likewise, gift_wrap is now a child of item/stack/package_wrap having the same functionality as cargo's wrapper. Now that their behavior is unified, each of these two items can do more than they did before.

Also made it that you require a sharp item to unwrap those items.

Spookerton commented on code/game/objects/items.dm:
```diff
@@ -289,6 +289,7 @@
 				S.handle_item_insertion(src)
 
 /obj/item/use_on(obj/target, mob/user)
+	SHOULD_CALL_PARENT(TRUE)
 	if (istype(target, /obj/item/clothing))
 		var/obj/item/clothing/clothes = target
 		if (clothes.attempt_store_item(src, user))
```
Oh dear. This should really be implemented as a receiving event on /clothing instead. use_* events should try to avoid demanding parent calls, but most especially on the base implementation.

emmanuelbassil replied:
Will do

emmanuelbassil replied:
Fixed, pending the checks to make sure there's no use_tool() clothing override not calling parent.

emmanuelbassil commented:
Moved webbing storage and accessory attachment to use_tool.
Also, while making use_tool on boots call parent, made it that any sharp item can now clip boots; not only scalpel/wirecutter.

MuckerMayhem commented on code/game/objects/items/weapons/gift_wrappaper.dm:
```diff
@@ -153,71 +153,91 @@
+/obj/item/wrapping_paper/use_tool(obj/item/tool, mob/user)
+	if (!istype(tool))
+		return FALSE
+	if (!(locate(/obj/structure/table, loc) ))
+		to_chat(user, SPAN_WARNING("You must put \the [src] on a table in order to wrap \the [tool]!"))
+		return TRUE
```
Why do items need to be on tables specifically? I'd say it's just as well someone could wrap things on the floor or a table.

emmanuelbassil replied:
Fair point. Will change. Will make it check being on a turf though to avoid wrapping stuff in storage

MuckerMayhem commented on code/modules/clothing/_clothing.dm:
```diff
@@ -205,6 +205,16 @@
+		return FALSE
+	if (attempt_attach_accessory(tool, user))
+		return TRUE
+	if (attempt_store_item(tool, user))
+		return TRUE
+	else return ..()
```
```suggestion
	if (attempt_store_item(tool, user))
		return TRUE
	return ..()
```

emmanuelbassil replied:
Fixed

emmanuelbassil commented:
Fixed requests. Also fixed being able to wrap a sharp item without using ANOTHER sharp item to cut the wrapping paper. It now checks that the offhand item is sharp.

SierraKomodo reviewed (changes requested):
Per discord, needs updating to rebase and update attackby counts

emmanuelbassil committed:
Improves Giftwrapping

Fixes

Auto stash before rebase of "refs/heads/AssortedImp"

Fixes

Fixes

Auto stash before rebase of "refs/heads/AssortedImp"

Auto stash before rebase of "refs/heads/AssortedImp"

Auto stash before rebase of "refs/heads/AssortedImp"

Auto stash before rebase of "refs/heads/AssortedImp"

Fixes

Auto stash before rebase of "refs/heads/AssortedImp"

emmanuelbassil commented:
Fixed the check-paths file. Ready to merge

SierraKomodo commented on code/game/objects/items/weapons/gift_wrappaper.dm:
```diff
@@ -153,71 +153,85 @@
-		to_chat(user, SPAN_WARNING("The object is FAR too large!"))
-	return
+/obj/item/wrapping_paper/use_tool(obj/item/tool, mob/user)
+	if (!istype(tool))
+		return FALSE
+	if (!isturf(loc))
```
This check shouldn't be necessary.

SierraKomodo commented on code/game/objects/items/weapons/gift_wrappaper.dm:
```diff
@@ -153,71 +153,85 @@
+
+	var/a_used = tool.get_storage_cost()
+	if (a_used == ITEM_SIZE_NO_CONTAINER)
+		to_chat(user, SPAN_WARNING("You can't wrap that!"))
```
```suggestion
		to_chat(user, SPAN_WARNING("You can't wrap \the [tool]!"))
```

SierraKomodo commented on code/game/objects/items/weapons/gift_wrappaper.dm:
```diff
@@ -153,71 +153,85 @@
+		to_chat(user, SPAN_WARNING("You can't wrap that!"))
+		return TRUE
+	if (amount < a_used)
+		to_chat(user, SPAN_WARNING("You need more paper!"))
```
```suggestion
		to_chat(user, SPAN_WARNING("\The [src] needs more paper to wrap \the [tool]!"))
```

SierraKomodo commented on code/game/objects/items/weapons/gift_wrappaper.dm:
```diff
@@ -153,71 +153,85 @@
+/obj/item/wrapping_paper/use_on(obj/item/target, mob/user)
+	if (!isitem(target))
+		return FALSE
+	if (!isturf(loc))
+		to_chat(user, SPAN_WARNING("You must put \the [src] on a surface in order to wrap \the [target]!"))
+		return TRUE
 
```
Technically speaking, this condition will always pass. I assume that's intentional though.

SierraKomodo commented on code/game/objects/items/weapons/gift_wrappaper.dm:
```diff
@@ -153,71 +153,85 @@
-	else
-		to_chat(user, SPAN_WARNING("The object is FAR too large!"))
-	return
+/obj/item/wrapping_paper/use_tool(obj/item/tool, mob/user)
```
Looking at the checks for these interactions - Is there a specific reason package wrapping can't also be done through `use_on()` alongside `use_tool()`? Of particular note is the fact crates should be wrappable for shipping via cargo disposals.

SierraKomodo commented on code/modules/clothing/shoes/miscellaneous.dm:
```diff
@@ -230,16 +230,21 @@
-		playsound(src.loc, 'sound/items/Wirecutter.ogg', 100, 1)
-		user.visible_message(SPAN_WARNING("\The [user] modifies \the [src] with \the [W]."),SPAN_WARNING("You modify \the [src] with \the [W]."))
-		cut_clogs()
+	if (!isitem(W))
+		return FALSE
+	if (!is_sharp(W))
```
Unnecessary check

SierraKomodo commented on code/modules/clothing/_clothing.dm:
```diff
@@ -207,6 +207,16 @@
 
+/obj/item/clothing/use_tool(obj/item/tool, mob/living/user, list/click_params)
+	SHOULD_CALL_PARENT(TRUE)
+	if (!isitem(tool))
+		return FALSE
+	if (attempt_attach_accessory(tool, user))
```
Unnecessary check

emmanuelbassil commented:
Code ready for review; PR still needs some sprites before being ready to merge.
Made gift wrapping and cargo wrapping behave exactly the same way, and they're both children of stacks now.

emmanuelbassil commented:
This is done and ready to go. Will change original comment reflecting changes.

SierraKomodo commented on code/datums/supplypacks/nonessent.dm:
```diff
@@ -23,7 +23,7 @@
 	/obj/item/reagent_containers/glass/paint/white,
 	/obj/item/contraband/poster,
-	/obj/item/wrapping_paper = 3)
+	/obj/item/stack/package_wrap/gift_wrap = 3)
```
To make this a touch more consistant with the cargo wrapper - One gift wrap here, and then a separate orderable pack that contains 3, like the supply wrap pack that can be ordered.

```suggestion
	/obj/item/stack/package_wrap/gift_wrap)
```

emmanuelbassil replied:
Done

SierraKomodo commented on code/game/objects/items/weapons/wrapping_paper.dm:
```diff
@@ -0,0 +1,162 @@
+	singular_name = "sheet"
+	amount = 25
+	max_amount = 25
+	w_class = ITEM_SIZE_NORMAL
```
This value is identical to parent.

```suggestion
```

SierraKomodo commented on code/game/objects/items/weapons/wrapping_paper.dm:
```diff
@@ -0,0 +1,162 @@
+/obj/item/stack/package_wrap/cargo_wrap
+	name = "package wrapper"
+	desc = "Heavy duty brown paper used to wrap packages to protect them during shipping."
+	singular_name = "sheet"
```
Identical to parent.


```suggestion
```

SierraKomodo commented on code/game/objects/items/weapons/wrapping_paper.dm:
```diff
@@ -0,0 +1,162 @@
+/obj/item/stack/package_wrap/cargo_wrap/cyborg
+	name = "package wrapper synthesizer"
+	gender = NEUTER
+	matter = null
```
Identical to parent.

```suggestion
```

SierraKomodo commented on code/game/objects/items/weapons/wrapping_paper.dm:
```diff
@@ -0,0 +1,162 @@
+	throw_speed = 4
+	throw_range = 5
+
+///Wrapping Procs
```
Comment is somewhat unnecessary. The proc name already implies it's the wrapping proc.

```suggestion
```

SierraKomodo commented on code/game/objects/items/weapons/wrapping_paper.dm:
```diff
@@ -0,0 +1,162 @@
+		parcel.wrapped.forceMove(parcel)
+		parcel.add_fingerprint(user)
+		amount_used = istype(target, /obj/structure/closet/crate) ? ITEM_SIZE_NORMAL : ITEM_SIZE_LARGE
+	if (istype(target, /obj/item))
```
```suggestion
	else if (istype(target, /obj/item))
```

SierraKomodo commented on code/game/objects/items/weapons/wrapping_paper.dm:
```diff
@@ -0,0 +1,162 @@
+	user.visible_message("\The [user] wraps \the [target] with \the [src].",\
+		SPAN_NOTICE("You wrap \the [target], leaving [get_amount()] units of paper on \the [src]."),\
+		"You hear someone taping paper around a small object.")
+
```
```suggestion
	user.visible_message(
        SPAN_NOTICE("\The [user] wraps \the [target] with [src.get_vague_name(true)]."),
		SPAN_NOTICE("You wrap \the [target] with [src.get_exact_name(amount_used)]."),
		"You hear someone taping paper around a small object."
    )

```

SierraKomodo commented on code/game/objects/items/weapons/wrapping_paper.dm:
```diff
@@ -0,0 +1,162 @@
+		return TRUE
+	var/amount = get_amount()
+	if (amount < 1)
+		to_chat(user, SPAN_WARNING("\The [src] ran out of paper."))
```
```suggestion
		to_chat(user, SPAN_WARNING("\The [src] is out of [plural_name]."))
```

SierraKomodo commented on code/game/objects/items/weapons/wrapping_paper.dm:
```diff
@@ -0,0 +1,162 @@
+		var/obj/item/target = object
+		var/a_used = target.get_storage_cost()
+		if (a_used == ITEM_SIZE_NO_CONTAINER || a_used == ITEM_SIZE_GARGANTUAN)
+			to_chat(user, SPAN_WARNING("You can't wrap \the [target]!"))
```
```suggestion
			to_chat(user, SPAN_WARNING("\The [target] is too big to wrap!"))
```

SierraKomodo commented on code/game/objects/items/weapons/wrapping_paper.dm:
```diff
@@ -0,0 +1,162 @@
+			to_chat(user, SPAN_WARNING("You must put down \the [target] in order to wrap it."))
+			return TRUE
+		if (amount < a_used)
+			to_chat(user, SPAN_WARNING("\The [src] needs more paper to wrap \the [target]!"))
```
```suggestion
			USE_FEEDBACK_STACK_NOT_ENOUGH(src, a_used, "to wrap \the [target]!")
```

SierraKomodo commented on code/game/objects/items/weapons/wrapping_paper.dm:
```diff
@@ -0,0 +1,162 @@
+		if (amount < a_used)
+			to_chat(user, SPAN_WARNING("\The [src] needs more paper to wrap \the [target]!"))
+			return TRUE
+		if (!do_after(user, target.w_class SECONDS, target, DO_PUBLIC_UNIQUE))
```
```suggestion
		if (!do_after(user, target.w_class SECONDS, target, DO_PUBLIC_UNIQUE) || !user.use_sanity_check(target, src))
```

SierraKomodo replied:
You also need to add a visible message before the do_after since it has a publically visible timer bar, to provide context for that bar.

SierraKomodo commented on code/game/objects/items/weapons/wrapping_paper.dm:
```diff
@@ -0,0 +1,162 @@
+		else
+			item_size = ITEM_SIZE_LARGE
+		if (amount < item_size)
+			to_chat(user, SPAN_WARNING("\The [src] needs more paper to wrap \the [target]!"))
```
```suggestion
			USE_FEEDBACK_STACK_NOT_ENOUGH(src, item_size, "to wrap \the [target]!")
```

SierraKomodo commented on code/game/objects/items/weapons/wrapping_paper.dm:
```diff
@@ -0,0 +1,162 @@
+			to_chat(user, SPAN_WARNING("\The [src] needs more paper to wrap \the [target]!"))
+			return TRUE
+		if (target.opened)
+			to_chat(user, SPAN_WARNING("You need to close \the [target]."))
```
```suggestion
			to_chat(user, SPAN_WARNING("You need to close \the [target] before you can wrap it."))
```

SierraKomodo commented on code/game/objects/items/weapons/wrapping_paper.dm:
```diff
@@ -0,0 +1,162 @@
+		if (target.opened)
+			to_chat(user, SPAN_WARNING("You need to close \the [target]."))
+			return TRUE
+		if (!do_after(user, item_size SECONDS, target, DO_PUBLIC_UNIQUE))
```
```suggestion
		if (!do_after(user, item_size SECONDS, target, DO_PUBLIC_UNIQUE) || !user.use_sanity_check(target, src))
```

SierraKomodo replied:
This also needs a visible message.

SierraKomodo commented on code/game/objects/items/weapons/wrapping_paper.dm:
```diff
@@ -0,0 +1,162 @@
+	var/a_used = BASE_STORAGE_COST(ITEM_SIZE_LARGE) //get_storage_cost() does not work on mobs, will reproduce same logic here.
+
+	if (get_amount() < a_used)
+		to_chat(user, SPAN_WARNING("You need more paper."))
```
```suggestion
		USE_FEEDBACK_STACK_NOT_ENOUGH(src, a_used, "to wrap \the [target]!")
```

SierraKomodo commented on code/game/objects/items/weapons/wrapping_paper.dm:
```diff
@@ -0,0 +1,162 @@
+		to_chat(user, SPAN_WARNING("You need more paper."))
+		return TRUE
+	if (!H.has_danger_grab(user))
+		to_chat(user, "You need to have a firm grip on \the [target] in order to wrap them.")
```
```suggestion
		to_chat(user, SPAN_WARNING("You need to have a firm grip on \the [target] in order to wrap them."))
```

SierraKomodo commented on code/game/objects/items/weapons/wrapping_paper.dm:
```diff
@@ -0,0 +1,162 @@
+	if (!H.has_danger_grab(user))
+		to_chat(user, "You need to have a firm grip on \the [target] in order to wrap them.")
+		return TRUE
+	if (!do_after(user, ITEM_SIZE_LARGE SECONDS, target, DO_PUBLIC_UNIQUE) || !H.has_danger_grab(user))
```
```suggestion
	if (!do_after(user, ITEM_SIZE_LARGE SECONDS, target, DO_PUBLIC_UNIQUE) || !H.has_danger_grab(user) || !user.use_sanity_check(H, src))
```

SierraKomodo replied:
Visible message here too.

SierraKomodo commented on code/game/objects/items/weapons/wrapping_paper.dm:
```diff
@@ -0,0 +1,162 @@
+		H.client.eye = present
+
+	if (user == target)
+		user.visible_message(SPAN_DANGER("\The [user] has gift-wrapped themselves, leaving [get_amount()] units of paper on \the [src]."))
```
```suggestion
		user.visible_message(
            SPAN_DANGER("\The [user] wraps themselves with [get_vague_name(true)].),
            SPAN_DANGER("You wrap yourself with [get_exact_name(a_used)].")
        )
```

SierraKomodo commented on code/game/objects/items/weapons/wrapping_paper.dm:
```diff
@@ -0,0 +1,162 @@
+	if (user == target)
+		user.visible_message(SPAN_DANGER("\The [user] has gift-wrapped themselves, leaving [get_amount()] units of paper on \the [src]."))
+	else
+		user.visible_message(SPAN_DANGER("\The [user] has gift-wrapped \the [target], leaving [get_amount()] units of paper on \the [src]."))
```
```suggestion
		user.visible_message(
			SPAN_DANGER("\The [user] wraps \the [target] with [get_vague_name(true)]."),
			SPAN_DANGER("You wrap \the [target] with [get_exact_name(a_used)].")
		)
```

SierraKomodo commented on code/game/objects/random/random.dm:
```diff
@@ -149,7 +149,7 @@
 	return list(/obj/random/powercell = 3,
 				/obj/random/technology_scanner = 2,
-				/obj/item/stack/package_wrap/twenty_five = 1,
+				/obj/item/stack/package_wrap/cargo_wrap = 1,
```
Might be worth adding gift wrap here too.

SierraKomodo commented on code/modules/recycling/sortingmachinery.dm:
```diff
@@ -14,6 +14,14 @@
+	..(mapload)
+	SetName("large [package_type]")
+	desc = name
+	if (istype(parcel, /obj/structure/closet/crate))
+		icon_state = "[package_type]crate"
+	else
+		icon_state = "[package_type]closet"
 
```
icon_state changes should occur in `on_update_icon()`.


```suggestion
    queue_icon_update()

/obj/structure/bigDelivery/on_update_icon()
	if (istype(wrapped, /obj/structure/closet/crate))
		icon_state = "[package_type]crate"
	else
		icon_state = "[package_type]closet"

```

SierraKomodo commented on code/modules/recycling/sortingmachinery.dm:
```diff
@@ -14,6 +14,14 @@
 	var/label_x
 	var/tag_x
 
+/obj/structure/bigDelivery/Initialize(mapload, obj/structure/closet/parcel, package_type)
+	..(mapload)
+	SetName("large [package_type]")
+	desc = name
+	if (istype(parcel, /obj/structure/closet/crate))
```
Also, I'm not seeing where `wrapped` is set to `parcel` here, which I'm assuming it needs to be.

emmanuelbassil commented:
Comments implemented.
To do in a separate PR when I feel like touching this again:
1) Add sounds for wrapping/unwrapping
2) Allow taggers/pens to interact with wrapped mobs like they can with wrapped crates/closets/items
3) Give labelers the same functionality as pens in terms of labeling wrapped stuff.

Who knew such a small item can take so much work; holy shit.

SierraKomodo reviewed (changes requested):
Some notes from testing:
- Clicking an open locker or crate should put the item in the locker, instead of failing to wrap.
- Clicking a stack with another stack of the same type should combine the stacks instead of wrapping the stack in paper (I believe this is covered in parent behaviour).
- Opening a wrapped item that's on the floor places the item in your hand instead of on the turf the package was originally on.

emmanuelbassil committed:
Makes gift_wrapper part of stacks.

Fixes

Fixes

Auto stash before rebase of "refs/heads/AssortedImp"

Auto stash before rebase of "refs/heads/AssortedImp"

Fixes

Auto stash before rebase of "refs/heads/AssortedImp"

Auto stash before rebase of "refs/heads/AssortedImp"

Auto stash before rebase of "refs/heads/AssortedImp"

Auto stash before rebase of "refs/heads/AssortedImp"

Auto stash before rebase of "refs/heads/AssortedImp"

Auto stash before rebase of "refs/heads/AssortedImp"

Auto stash before rebase of "refs/heads/AssortedImp"

Auto stash before rebase of "refs/heads/AssortedImp"

Auto stash before rebase of "refs/heads/AssortedImp"

Auto stash before rebase of "refs/heads/AssortedImp"

Fixes

Auto stash before rebase of "refs/heads/AssortedImp"

Auto stash before rebase of "refs/heads/Squish"

emmanuelbassil commented:
Requests implemented.
Of note; moving this use_on to use_after in the next PR this depends on will allow me to remove these checks (which I will).
Another benefit of having the override stuff split into before/after. Currently /stack/package_wrap requires the ITEM_TRY_TO_ATTACK flag to wrap other people; hence I can't have use_on run before the storage/base stack procs.
